Add LinearSystem and TimeVaryingLinearSystem specializations to Direct Transcription

jadecastro opened the pull request:
Additionally:
- Introduce the TimeVaryingLinearSystem class (and unit tests).
- Update the unit tests for DirTran, namely for the symbolic constraints, and extend to encompass the new constructors.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7164)
<!-- Reviewable:end -->

jadecastro commented:
+@russtedrake for feature review, please.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-KvU_6Ye8TQDNNLxy0Hg:b-dmufd6)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Checkpoint

---

Reviewed 1 of 8 files at r1.
Review status: 1 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-KvVVw8UULNnja36M8zy:bid65y5) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
> ```cpp
> 
>   for (int i = 0; i < N() - 1; i++) {
>     Eigen::MatrixXd Aeq =
> ```

For readability, you could change the code in the for loop as
```C++
AddLinearEqualityConstraint(state(i+1) == system->A() * state(i) + system->B() * input(i));
```

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVS7dd3pLV7bstH2Rn:-KvVS7deBWF2c9GqBjES:b-z8f819) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.h#L108)):*
> ```objc
>                                      const Context<double>& context);
> 
>     // Constrain the final input to match the penultimate, otherwise the final
> ```

Check indentation?

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVSPt7CmvuV5Vmc9rL:-KvVSPt7CmvuV5Vmc9rM:bxzguq3) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
> ```objc
>   // (Note that it might be more ideal to have less decision variables allocated
>   // for this specific case, but this is a reasonable work-around).
>   void AddInputConstraintAtFinalTime();
> ```

The function name is not very indicative, of what it actually does. Maybe you mean `ConstrainFinalStateAsEquilibriumPoint`?

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 114 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVSsag5-FiRuyZB468:-KvVSsag5-FiRuyZB469:bdqjpph) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.h#L114)):*
> ```objc
>   void AddInputConstraintAtFinalTime();
> 
>   void ValidateSystem(const System<double>& system,
> ```

This function needs documentation. And maybe move the function definition to the `.cc` file?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-KvVXxGFzFn18lQ5Xjk7:b13s2ul)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-KvYxOhGSaMcdPYCdBss:bgmnrqh) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

For readability, you could change the code in the for loop as
```C++
AddLinearEqualityConstraint(state(i+1) == system->A() * state(i) + system->B() * input(i));
```
</blockquote></details>

@hongkai-dai that unfortunately does not work.  I get the following:
```
drake/systems/trajectory_optimization/direct_transcription.cc:137:20: note: in instantiation of function template specialization 'drake::symbolic::operator==<Eigen::VectorBlock<const Eigen::Matrix<drake::symbolic::Variable, -1, 1, 0, -1, 1>, -1>, Eigen::CwiseBinaryOp<Eigen::internal::scalar_sum_op<drake::symbolic::Expression, drake::symbolic::Expression>, const Eigen::Product<Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Block<const Eigen::Matrix<drake::symbolic::Variable, -1, 1, 0, -1, 1>, -1, 1, false>, 0>, const Eigen::Product<Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Block<const Eigen::Matrix<drake::symbolic::Variable, -1, 1, 0, -1, 1>, -1, 1, false>, 0> > >' requested here
        state(i+1) ==
                   ^
external/eigen/Eigen/src/Core/GenericPacketMath.h:222:1: note: candidate function [with Packet = drake::symbolic::Variable] not viable: no known conversion from 'drake::symbolic::Expression' to 'const typename unpacket_traits<Variable>::type' (aka 'const drake::symbolic::Variable') for 1st argument
pset1(const typename unpacket_traits<Packet>::type& a) { return a; }

```

This happens even if I first convert the system into symbolic form.  The alternative is to just do an `AddSymbolicDynamicConstraints()` on the system, which does essentially the same thing, but with a bit more overhead.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVS7dd3pLV7bstH2Rn:-KvY1W71OjNOTj60dA-v:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.h#L108)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Check indentation?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVSPt7CmvuV5Vmc9rL:-KvY1q3rP3wnJzjpkgNp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

The function name is not very indicative, of what it actually does. Maybe you mean `ConstrainFinalStateAsEquilibriumPoint`?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 114 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVSsag5-FiRuyZB468:-KvY3NdMij_SussaxLyS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.h#L114)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This function needs documentation. And maybe move the function definition to the `.cc` file?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-KvZ4YCYZQQUQmZLtg8k:b-v7e2oi) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

@hongkai-dai that unfortunately does not work.  I get the following:
```
drake/systems/trajectory_optimization/direct_transcription.cc:137:20: note: in instantiation of function template specialization 'drake::symbolic::operator==<Eigen::VectorBlock<const Eigen::Matrix<drake::symbolic::Variable, -1, 1, 0, -1, 1>, -1>, Eigen::CwiseBinaryOp<Eigen::internal::scalar_sum_op<drake::symbolic::Expression, drake::symbolic::Expression>, const Eigen::Product<Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Block<const Eigen::Matrix<drake::symbolic::Variable, -1, 1, 0, -1, 1>, -1, 1, false>, 0>, const Eigen::Product<Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Block<const Eigen::Matrix<drake::symbolic::Variable, -1, 1, 0, -1, 1>, -1, 1, false>, 0> > >' requested here
        state(i+1) ==
                   ^
external/eigen/Eigen/src/Core/GenericPacketMath.h:222:1: note: candidate function [with Packet = drake::symbolic::Variable] not viable: no known conversion from 'drake::symbolic::Expression' to 'const typename unpacket_traits<Variable>::type' (aka 'const drake::symbolic::Variable') for 1st argument
pset1(const typename unpacket_traits<Packet>::type& a) { return a; }

```

This happens even if I first convert the system into symbolic form.  The alternative is to just do an `AddSymbolicDynamicConstraints()` on the system, which does essentially the same thing, but with a bit more overhead.
</blockquote></details>

Interesting. I'll check the details this evening. I'll open a separate PR if I make it work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-KvZ6sLS7fayV8O08QMm:bwrc1pc) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Interesting. I'll check the details this evening. I'll open a separate PR if I make it work.
</blockquote></details>

I guess we can try
```C++
AddLinearEqualityConstraint(state(i+1).cast<symbolic::Expression>()== system->A() * state(i) + system->B() * input(i));
```
Sorry I have not tested this yet. I am re-installing my computer without a working drake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
BTW @hongkai-dai - would you like to feature review, since you've dug this far already?

---

Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-Kvb1jG2s4dALQQYQbN-:b-2maedl) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I guess we can try
```C++
AddLinearEqualityConstraint(state(i+1).cast<symbolic::Expression>()== system->A() * state(i) + system->B() * input(i));
```
Sorry I have not tested this yet. I am re-installing my computer without a working drake.
</blockquote></details>

Sorry, it doesn't work either. I've dropped a `TODO` in here, until we can resolve. Does this work? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-Kvb2wZ3DMXXlnaSngp5:bq4ak84)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-KvbdfTrV9u6IFlJgptW:b-ib8k3h) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Sorry, it doesn't work either. I've dropped a `TODO` in here, until we can resolve. Does this work? 
</blockquote></details>

I agree that `TODO` is good enough. I'm checking it now. If I find a solution in time, I'll post it here. Otherwise, another PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-Kvbkde1oWNLoDqnY6N4:bg1dww1) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I agree that `TODO` is good enough. I'm checking it now. If I find a solution in time, I'll post it here. Otherwise, another PR.
</blockquote></details>

Hmm, ok. I was able to get the following to work (in retrospect, I should have tried this first):
```
    AddLinearEqualityConstraint(
        state(i+1).cast<symbolic::Expression>() ==
        system->A() * state(i).cast<symbolic::Expression>() +
        system->B() * input(i).cast<symbolic::Expression>());
```
I think this makes things a bit more readable?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Yes, I would be happy to review it! +@hongkai-dai for feature review.

---

Reviewed 1 of 2 files at r2.
Review status: 1 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvVVw8UULNnja36M8zx:-Kvc4e_YLx4L1UXlqB9Z:bx08uj2) ([raw file](https://github.com/robotlocomotion/drake/blob/5c1f3e729d8f6d91328bf993fa9d7d13a30a74bd/drake/systems/trajectory_optimization/direct_transcription.cc#L129)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Hmm, ok. I was able to get the following to work (in retrospect, I should have tried this first):
```
    AddLinearEqualityConstraint(
        state(i+1).cast<symbolic::Expression>() ==
        system->A() * state(i).cast<symbolic::Expression>() +
        system->B() * input(i).cast<symbolic::Expression>());
```
I think this makes things a bit more readable?
</blockquote></details>

Ah you are absolutely right, I forgot to cast state and input to symbolic expression on the right-hand side. Yes now the code looks a lot more clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-Kvc4WJJbwXTW0YpHvYM:b5x4su8)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Checkpoint, sorry for the delay.

---

Reviewed 2 of 2 files at r3.
Review status: 3 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 277 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd8yQfrkx8y50dikXJ:-Kvd8yQfrkx8y50dikXK:b9ecf3j) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.cc#L277)):*
> ```cpp
> void DirectTranscription::ConstrainFinalStateAsEquilibriumPoint() {
>   if (num_inputs() > 0) {
>     AddLinearConstraint(input(N() - 2) == input(N() - 1));
> ```

I realize what this function does is not to constrain the final state, but just the final input. So maybe rename this function as `ConstraintFinalInputSameAsTheSecondLast`?

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd9KfB44EKX-CNgj9E:-Kvd9KfB44EKX-CNgj9F:b-pk3e7q) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
> ```objc
>   // (Note that it might be more ideal to have less decision variables allocated
>   // for this specific case, but this is a reasonable work-around).
>   void ConstrainFinalStateAsEquilibriumPoint();
> ```

Why we have to constrain the final input? Without that constraint, this is still a valid optimization program. The final input just does not matter in the optimization.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 102 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvfDiDS-gxRaBotRflS:-KvfDiDS-gxRaBotRflT:bc94cz) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L102)):*
> ```cpp
> 
> std::unique_ptr<AffineSystem<double>> MakeAffineSystem(const double time_step) {
>   // x[n+1] = 1.
> ```

I would suggest to assign the affine system with non-zero A, B, C, D matrices. With A, B, C, D all zero, this is a very special form of affine system. It is probably better to test a more general case.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvfCMuOxM4BeKCw-Vv4:-KvfCMuOxM4BeKCw-Vv5:b-h9csy2) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L162)):*
> ```cpp
>   // Constructor should add dynamic constraints, and these should have been
>   // added as linear equality constraints. (There is also one linear
>   // equality constraint for the control input).
> ```

Is the linear equality constraint for the control input also checked in the test?
There should be `kNumSampleTimes - 1` linear equality constraints in the form
```
state(i+1) - A * state(i) - B * input(i) - x0 = 0
```
And one linear equality constraint in the form
```
input(kNumSampleTimes - 1) == input(kNumSampleTimes - 2)
```
So in total should be `kNumSampleTime` linear equality constraints. What do I miss?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-KvfG80Gpj4ttxCU_Vzo:b5eth1x)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Thanks for the comments, I've addressed all but one.

---

Review status: 3 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 277 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd8yQfrkx8y50dikXJ:-KvgtP1rTJ148xndRkpU:btieqxp) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.cc#L277)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I realize what this function does is not to constrain the final state, but just the final input. So maybe rename this function as `ConstraintFinalInputSameAsTheSecondLast`?
</blockquote></details>

You're right. I've changed the name. 

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd9KfB44EKX-CNgj9E:-KvgyW6ckYMmo2zwgeKw:b-uvftfn) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Why we have to constrain the final input? Without that constraint, this is still a valid optimization program. The final input just does not matter in the optimization.
</blockquote></details>

My thinking (please correct me if I'm wrong) is that the final input at time N does not influence the trajectory, yet it remains among the decision variables (because we need a trajectory of the form {x(k), u(k)}ₖ₌₀ᵏ⁼ᴺ). So, you are right that it does not play a role in the optimization problem. But since the user is free to not to choose a cost function that penalizes control effort, then we do not in general want that value to become unbounded.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 102 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvfDiDS-gxRaBotRflS:-KvhCNxuvAzX12oZR3TO:b-6nwwtl) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L102)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I would suggest to assign the affine system with non-zero A, B, C, D matrices. With A, B, C, D all zero, this is a very special form of affine system. It is probably better to test a more general case.
</blockquote></details>

Good point. I'll work up a system that populates A, B, C, and D.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvfCMuOxM4BeKCw-Vv4:-KvhCDU8UGZ0r9-o-3fC:bj0dklv) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L162)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Is the linear equality constraint for the control input also checked in the test?
There should be `kNumSampleTimes - 1` linear equality constraints in the form
```
state(i+1) - A * state(i) - B * input(i) - x0 = 0
```
And one linear equality constraint in the form
```
input(kNumSampleTimes - 1) == input(kNumSampleTimes - 2)
```
So in total should be `kNumSampleTime` linear equality constraints. What do I miss?
</blockquote></details>

There's no input for this system, hence no additional constraint on the input at the final time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-Kvgt8ybZeM-4Wjrpmk9:b71279i)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 3 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd9KfB44EKX-CNgj9E:-KvhDkEhl_UIKxToki9l:bkybeie) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

My thinking (please correct me if I'm wrong) is that the final input at time N does not influence the trajectory, yet it remains among the decision variables (because we need a trajectory of the form {x(k), u(k)}ₖ₌₀ᵏ⁼ᴺ). So, you are right that it does not play a role in the optimization problem. But since the user is free to not to choose a cost function that penalizes control effort, then we do not in general want that value to become unbounded.
</blockquote></details>

I see the point to make each decision variable bounded. But it looks to me that another solution is to remove `u(N)` from the decision variables. In the current implementation by setting u(N) == u(N-1), the optimization could become infeasible, if the user naively add a constraint u(N) = 0, when he wants the last state to be an equilibrium point.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvfCMuOxM4BeKCw-Vv4:-KvhHJ4PV-8ooPd6jUeE:b-ifdz16) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L162)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

There's no input for this system, hence no additional constraint on the input at the final time.
</blockquote></details>

I see, thanks for the explanation! I forgot that B and D are 0.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
-@RussTedrake feature review (though comments most welcome!)

---

Review status: 3 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd9KfB44EKX-CNgj9E:-KvhJNxRlpot68gl2q6M:b-2b0p70) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I see the point to make each decision variable bounded. But it looks to me that another solution is to remove `u(N)` from the decision variables. In the current implementation by setting u(N) == u(N-1), the optimization could become infeasible, if the user naively add a constraint u(N) = 0, when he wants the last state to be an equilibrium point.
</blockquote></details>

Hmm. That's a good suggestion, but not entirely consistent with the design that's here already. Is there a way to remove a decision variable (resize the `VectorXDecisionVariable`) that has already been made by the MultipleShooting subclass (it's currently a `const` member field)? But I feel it might also be dangerous to give consumers of MultipleShooting the ability to remove decision variables. Or perhaps I could add a new constructor that creates N - 1 input decision variables.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-KvhCsMhDR-YlKO12siv:bm5aw65)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 3 of 8 files at r1.
Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/primitives/time_varying_linear_system.h, line 40 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvhS1PfUwkLG-wsw9kC:-KvhS1PfUwkLG-wsw9kD:b-za8jqc) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/primitives/time_varying_linear_system.h#L40)):*
> ```objc
>   }
> 
>   PiecewisePolynomialTrajectory A{PiecewisePolynomial<double>()};
> ```

This assumes that `A, B, C, D` are piecewise polynomials, which can be restrictive (for example, we cannot use sin/cos functions). I would suggest to rename the class as `LinearTimeVaryingPiecewisePolynomialData` or something  alike.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd9KfB44EKX-CNgj9E:-KvhcmejchiKVUHc5KR_:b-lo0n9c) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Hmm. That's a good suggestion, but not entirely consistent with the design that's here already. Is there a way to remove a decision variable (resize the `VectorXDecisionVariable`) that has already been made by the MultipleShooting subclass (it's currently a `const` member field)? But I feel it might also be dangerous to give consumers of MultipleShooting the ability to remove decision variables. Or perhaps I could add a new constructor that creates N - 1 input decision variables.
</blockquote></details>

I see, you are right. The decision variables are constructed by the parent class `MultipleShooting`, that you cannot control. Then I think there are two options.
1. In the constructor of `MultipleShooting`, pass in an argument `bool requires_final_input`, which defaults to true. When the boolen is true, `MultipleShooting` adds `N` inputs as decision variables; otherwise it only adds `N-1` inputs.
2. Leave `MultipleShooting` untouched. We can add constraint on `u(N)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Apologies for the delay; comments addressed - PTAL.

---

Review status: 0 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd9KfB44EKX-CNgj9E:-KvlyWC5PsCw3S9mTZ4n:bvmx4fv) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I see, you are right. The decision variables are constructed by the parent class `MultipleShooting`, that you cannot control. Then I think there are two options.
1. In the constructor of `MultipleShooting`, pass in an argument `bool requires_final_input`, which defaults to true. When the boolen is true, `MultipleShooting` adds `N` inputs as decision variables; otherwise it only adds `N-1` inputs.
2. Leave `MultipleShooting` untouched. We can add constraint on `u(N)`.
</blockquote></details>

I actually spent time digging into replacing `bool requires_final_input` to `MultipleShooting`.  It's a much nicer option, but in the end requires a somewhat involved reworking of `MultipleShooting`.  For instance, one question is - if we don't have a decision variable for `u` at time N, then does `AddConstraintToAllKnotPoints()` apply constraints from time 0....N-1?  Or do we apply 0...N and limit the semantics of the formula?  A similar thing applies to `AddFinalCost()`.  So, basically, I think I will do option 2 for now, and I've added a TODO. 

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 102 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvfDiDS-gxRaBotRflS:-KvlnsP82cgc1CP1rhTJ:bskxbh0) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L102)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Good point. I'll work up a system that populates A, B, C, and D.
</blockquote></details>

Done. PTAL. I've reworked both `DiscreteTimeSymbolicConstraintTest` and `AddRunningCostTest` that use the reconfigured example AffineSystem.

---

*[drake/systems/primitives/time_varying_linear_system.h, line 40 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KvhS1PfUwkLG-wsw9kC:-KvyY1AEY_B03qhYoxD4:be2cumw) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/primitives/time_varying_linear_system.h#L40)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This assumes that `A, B, C, D` are piecewise polynomials, which can be restrictive (for example, we cannot use sin/cos functions). I would suggest to rename the class as `LinearTimeVaryingPiecewisePolynomialData` or something  alike.
</blockquote></details>

Done. I've refactored this into a new system `PiecewisePolynomialSystem` that derives from a more generic base class that doesn't restrict the form of the matrices. 

(BTW, it looks like git didn't detect the filename change, so this one appears as deleted, when it's actually replaced with piecewise_polynomial_linear_system.h.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-Kvyfio-nxx3jbPE-zo4:buyi003)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
@hongkai-dai BTW, per your suggestion, I've also added a constructor to PiecewisePolynomialLinearSystem that takes in a vector of Eigen Matrices. 

---

Review status: 0 of 12 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-Kw05NOYeFdxD8C0z6o2:b-yj0bsp)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kvd9KfB44EKX-CNgj9E:-Kw3CI18PeRAQu1qh3w5:b-134ab9) ([raw file](https://github.com/robotlocomotion/drake/blob/321c4bbd18d281bbd7946b75997969294f4b0f90/drake/systems/trajectory_optimization/direct_transcription.h#L112)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

I actually spent time digging into replacing `bool requires_final_input` to `MultipleShooting`.  It's a much nicer option, but in the end requires a somewhat involved reworking of `MultipleShooting`.  For instance, one question is - if we don't have a decision variable for `u` at time N, then does `AddConstraintToAllKnotPoints()` apply constraints from time 0....N-1?  Or do we apply 0...N and limit the semantics of the formula?  A similar thing applies to `AddFinalCost()`.  So, basically, I think I will do option 2 for now, and I've added a TODO. 
</blockquote></details>

I see, that totally makes sense!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Almost <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. Just a few minor comments. +@sammy-tri  for platform review please.

---

Reviewed 9 of 15 files at r4.
Review status: 9 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/primitives/linear_system.h, line 141 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw5ug8Oh-mQ00PI-6Xp:-Kw5ug8Oh-mQ00PI-6Xq:blrqcuc) ([raw file](https://github.com/robotlocomotion/drake/blob/0f7f00ebda41380205d23a0825ce3da3578b05a4/drake/systems/primitives/linear_system.h#L141)):*
> ```objc
> 
>  private:
>   VectorX<T> f0(const T&) const final {
> ```

This is not a part of the public interface, but it could help the user and future maintainer to understand what `f0` and `y0` means, if we add a short description as documentation.

---

*[drake/systems/primitives/test/linear_system_test.cc, line 145 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw5se62MFmSmRYeMVpC:-Kw5se62MFmSmRYeMVpD:b-h1df20) ([raw file](https://github.com/robotlocomotion/drake/blob/0f7f00ebda41380205d23a0825ce3da3578b05a4/drake/systems/primitives/test/linear_system_test.cc#L145)):*
> ```cpp
>   EXPECT_EQ(sys.get_num_output_ports(), 1);
>   EXPECT_EQ(sys.get_num_input_ports(), 1);
>   EXPECT_NO_THROW(sys.A(0.));
> ```

Maybe also check the value of `A` matrix?
```C++
EXPECT_TRUE(CompareMatrices(sys.A(0), Eigen::Matrix2d::Identity()));
```

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 171 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw5vTcyD_CkFzXh0T_E:-Kw5vTcyD_CkFzXh0T_F:bisokh6) ([raw file](https://github.com/robotlocomotion/drake/blob/0f7f00ebda41380205d23a0825ce3da3578b05a4/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L171)):*
> ```cpp
>   for (int i = 0; i < (kNumSampleTimes - 1); i++) {
>     EXPECT_TRUE(CompareMatrices(
>         prog.EvalBindingAtSolution(dynamic_constraints[i]) -
> ```

I think it is more robust to check either
dynamics_constraint - lower_bound = x[n+1] - A * x[n] - B * u[n] - f0
or
dynamics_constraint - lower_bound = -(x[n+1] - A * x[n] - B * u[n] - f0)

In the MathematicalProgram, the user specifies the formula as
x[n+1] == A * x[n] + B * u[n] + f0
This formula can be converted to either
x[n+1] - A * x[n] - B * u[n] - f0 = 0
by moving the right-hand side to the left, or
A * x[n] + B * u[n] + f0 - x[n+1] = 0
by moving the left-hand side to the right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-Kw5uuUWS32shknlTYaX:b-xt86rl)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 6 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/primitives/linear_system.h, line 141 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw5ug8Oh-mQ00PI-6Xp:-Kw6vKgIWVRQWJNBP2H1:bm2alwc) ([raw file](https://github.com/robotlocomotion/drake/blob/0f7f00ebda41380205d23a0825ce3da3578b05a4/drake/systems/primitives/linear_system.h#L141)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

This is not a part of the public interface, but it could help the user and future maintainer to understand what `f0` and `y0` means, if we add a short description as documentation.
</blockquote></details>

Done. (Description added)

---

*[drake/systems/primitives/test/linear_system_test.cc, line 145 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw5se62MFmSmRYeMVpC:-Kw6tlSgWHHxbDZHHsnM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0f7f00ebda41380205d23a0825ce3da3578b05a4/drake/systems/primitives/test/linear_system_test.cc#L145)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Maybe also check the value of `A` matrix?
```C++
EXPECT_TRUE(CompareMatrices(sys.A(0), Eigen::Matrix2d::Identity()));
```
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 171 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw5vTcyD_CkFzXh0T_E:-Kw71VkDk-tcPGCMz4tT:bx53aa7) ([raw file](https://github.com/robotlocomotion/drake/blob/0f7f00ebda41380205d23a0825ce3da3578b05a4/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L171)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think it is more robust to check either
dynamics_constraint - lower_bound = x[n+1] - A * x[n] - B * u[n] - f0
or
dynamics_constraint - lower_bound = -(x[n+1] - A * x[n] - B * u[n] - f0)

In the MathematicalProgram, the user specifies the formula as
x[n+1] == A * x[n] + B * u[n] + f0
This formula can be converted to either
x[n+1] - A * x[n] - B * u[n] - f0 = 0
by moving the right-hand side to the left, or
A * x[n] + B * u[n] + f0 - x[n+1] = 0
by moving the left-hand side to the right.
</blockquote></details>

Good suggestion -- thanks. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 6 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 183 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw73x_Pm3xyLyR5TwaK:-Kw73x_Pm3xyLyR5TwaL:b-mpn2mp) ([raw file](https://github.com/robotlocomotion/drake/blob/e456fd666c3315f64511d53cb15f34a2bd1d5517/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L183)):*
> ```cpp
>     EXPECT_TRUE(
>         CompareMatrices(dynamic_constraint_val, dynamic_constraint_expected) ||
>         CompareMatrices(-dynamic_constraint_val, -dynamic_constraint_expected));
> ```

I think the second row should be
CompareMatrices(dynamic_constraint_val, -dynamic_constraint_expected);
otherwise it is the same as the first line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 6 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 183 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-Kw73x_Pm3xyLyR5TwaK:-Kw74x20vIXKX8uxHP94:bxhd8t8) ([raw file](https://github.com/robotlocomotion/drake/blob/e456fd666c3315f64511d53cb15f34a2bd1d5517/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L183)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think the second row should be
CompareMatrices(dynamic_constraint_val, -dynamic_constraint_expected);
otherwise it is the same as the first line.
</blockquote></details>

Oops. I noticed that too right after I pushed. Fixed. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Sorry for not making it to this today.  Will review tomorrow.

---

Review status: 6 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-Kw7ZqobzOsZzqVCsqS8:b-53ftqf)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
First pass complete.

---

Reviewed 9 of 15 files at r4, 4 of 4 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/primitives/piecewise_polynomial_linear_system.h, line 64 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwAzX04BQmFNvJWHHxn:-KwAzX04BQmFNvJWHHxo:b-exrpgv) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/primitives/piecewise_polynomial_linear_system.h#L64)):*
> ```objc
> 
> // Return a vector of the given size, with result[i] == i * step.
> inline std::vector<double> vector_iota(int size, double step) {
> ```

Is this intended to be made part of the public API in `drake::systems`?  If not, consider wrapping it `namespace detail` (similar to `assert_if_is_constraint` in `mathematical_program.h`) or some other method of hiding it (static class method in the private section would be another way).

---

*[drake/systems/primitives/piecewise_polynomial_linear_system.h, line 96 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwB-AMyQDBs3ZwPgUK9:-KwB-AMyQDBs3ZwPgUKA:bx6bf8) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/primitives/piecewise_polynomial_linear_system.h#L96)):*
> ```objc
>   ///  time_period=0.0 to denote a continuous time system.  @default 0.0
>   ///
>   /// Subclasses must use the protected constructor, not this one.
> ```

This class is declared `final` so I suspect there aren't going to be any subclasses?  (also applies to the other constructors)

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 75 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwB2BdBD3O-x8CYLE6p:-KwB2BdCSCtmGwGfDjiJ:b-3yoim5) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/trajectory_optimization/direct_transcription.h#L75)):*
> ```objc
>   /// @param num_time_samples The number of knot points in the trajectory.
>   /// @throws std::runtime_error If the system is not discrete time (only).
>   DirectTranscription(const PiecewisePolynomialLinearSystem<double>* system,
> ```

BTW is there a reason why we specifically use `PiecewisePolynomialLinearSystem` here instead of the parent `TimeVaryingLinearSystem`?  (I actually don't know, so this is a bit of a curiosity question on my part)

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 272 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwB55-cvkAjX6u8I8cI:-KwB55-dfm51T8TkfDoM:b-b5iiib) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L272)):*
> ```cpp
>   DirectTranscription prog(&system, *context, kNumSampleTimes);
> 
>   // TODO(russt):  Uncomment this upon resolution of #6878.
> ```

Is this TODO applicable here?  I think the version of the constructor in use here actually does get the timestep from the system.  (this might also be true for the LinearSystem example above now?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-KwB5bdBk9eneHAaXq_Q:bvb1a6z)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 8 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/primitives/piecewise_polynomial_linear_system.h, line 64 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwAzX04BQmFNvJWHHxn:-KwBGnjTmjwMGfDUIOpc:b-ou6hvf) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/primitives/piecewise_polynomial_linear_system.h#L64)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Is this intended to be made part of the public API in `drake::systems`?  If not, consider wrapping it `namespace detail` (similar to `assert_if_is_constraint` in `mathematical_program.h`) or some other method of hiding it (static class method in the private section would be another way).
</blockquote></details>

Thanks for reminding me to deal properly with this function. I've made it a private member.

---

*[drake/systems/primitives/piecewise_polynomial_linear_system.h, line 96 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwB-AMyQDBs3ZwPgUK9:-KwBHK7msazoBbGB1zyu:b-njzfze) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/primitives/piecewise_polynomial_linear_system.h#L96)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

This class is declared `final` so I suspect there aren't going to be any subclasses?  (also applies to the other constructors)
</blockquote></details>

Removed, thanks.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 75 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwB2BdBD3O-x8CYLE6p:-KwBHPkB6L-LaqK_taeL:b-8ffewl) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/trajectory_optimization/direct_transcription.h#L75)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW is there a reason why we specifically use `PiecewisePolynomialLinearSystem` here instead of the parent `TimeVaryingLinearSystem`?  (I actually don't know, so this is a bit of a curiosity question on my part)
</blockquote></details>

It makes sense to base this constructor on the more general `TimeVaryingLinearSystem`, so I've changed it to take in that class instead of its specialization.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 272 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-KwB55-cvkAjX6u8I8cI:-KwBVqH7VGz_2fDCWwYR:b-q86r3y) ([raw file](https://github.com/robotlocomotion/drake/blob/1bd7b4c47c94206364f875a64e3b865bd988e6a8/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L272)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Is this TODO applicable here?  I think the version of the constructor in use here actually does get the timestep from the system.  (this might also be true for the LinearSystem example above now?)
</blockquote></details>

Good call. You're right. The version of the constructor used here (and in the LinearSystem test) does infer the time step from the system itself.  To make it work, I promoted `fixed_timestep()` to be a `public` member in `MultipleShooting` (it was `protected` before).  Please have a look.

The most general form of the constructor still does not do this, so the TODO should remain for those unit tests. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7164#-:-KwBruxkCbGbn-gUlGX5:bnfp4nl)*
<!-- Sent from Reviewable.io -->
